### PR TITLE
BUG: Handling of guess_bounds for circular coords

### PR DESCRIPTION
**Pull request by cpelley**
(1). **Fixes: `Coord.guess_bounds`, for correctly handling wraparound.**

Found the bug while performing an intersection (raised exception within `Cube._intersect_modulus`):

```
ValueError: coordinate's range greater than coordinate's unit's modulus
```

 (2). **Fixes `Coord.nearest_neighbour_index` for coords with descending bounds.**

---

**Note:**

This PR contains two fixes, as the existing 'integration' tests for `nearest_neighbour_index` draw on the bugged behaviour of the `guess_bounds` fixed in pnt1 above.  These tests have now been removed and new unit tests coded to replace them.

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1125,14 +1132,21 @@ def nearest_neighbour_index(self, point):
...
+            direction = 1 if self.bounds[0, 1] > self.bounds[0, 0] else -1
             bounds = bounds.copy()
             # sort the bounds cells by their centre values
             sort_inds = np.argsort(np.mean(bounds, axis=1))
             bounds = bounds[sort_inds]
             # replace all adjacent bounds with their averages
-            mid_bounds = 0.5 * (bounds[:-1, 1] + bounds[1:, 0])
-            bounds[:-1, 1] = mid_bounds
-            bounds[1:, 0] = mid_bounds
+            if direction is 1:
```
Comment (cpelley) [inline lib/iris/coords.py]: Sure

**Comment by cpelley** [comment]
Thanks @rhattersley, all ready, if your happy, I'll squash it.

**Comment by rhattersley** [comment]
> if your happy, I'll squash it.

Please do. Thanks.

**Comment by cpelley** [comment]
Squashed, cheers

**Comment by cpelley** [comment]
I want to check a couple of things before this goes in to be on the safe side (checking certain conditions are met when this branch is merged into master)

**Comment by cpelley** [comment]
Scratch the above comment, I should do that after this has been merged into 1.7.x
Ready

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1001,10 +1001,17 @@ def _guess_bounds(self, bound_position=0.5):
             raise ValueError('Coord already has bounds. Remove the bounds '
                              'before guessing new ones.')
 
-        diffs = np.diff(self.points)
-
-        diffs = np.insert(diffs, 0, diffs[0])
-        diffs = np.append(diffs, diffs[-1])
+        if getattr(self, 'circular', False) is True:
```
Comment (esc24) [inline lib/iris/coords.py]: Hi @cpelley. I'm :+1: overall, but would you mind dropping the `is True` (see PEP8)?

Comment (cpelley) [inline lib/iris/coords.py]: Sure

**Comment by cpelley** [comment]
Done, cheers @esc24 (and squashed).
Applogies for the PR journey, I'll take a more detailled comb before submitting in future.
Thanks both

**Pull request merged**